### PR TITLE
Schemas for UserClientMap and similar

### PR DESCRIPTION
**Pull request by pcapriotti**
Since `UserClientMap (Maybe Prekey)` is (and already was) a somewhat special case, this also wraps it in a `newtype` to give it a more specific non-overlapping `ToSchema` instance. The noise in the PR is mostly to follow this change.

### CODE REVIEW

**Review by jschaul — CHANGES_REQUESTED** [review]
Hm, shouldn't this change a golden JSON test? In the ToJSON, there was a `let key = Text.E.decodeLatin1 (toASCIIBytes u)` bit which disappeared and I don't have confidence to believe the output will now look exactly as before without a test demonstrating that this is indeed the case.
Also, the integration tests are changed at the same time as the implementation; how can we ensure the resulting jsons sent over the wire are compatible with the previous ways those looked like?

**Comment by akshaymankar** [comment]
I think that logic of translating user ids into string was doing pretty much the same thing as what the [`ToJSONKey` instance of `Id a`](https://github.com/wireapp/wire-server/blob/3795519231211b8876855860281229f3d741e939/libs/types-common/src/Data/Id.hs#L156-L157) does. Also, as we have golden tests we should be able to look at them and see if they are indeed testing with _realistic_ data or not. As this is just UUIDs, I think they are testing the right thing and correctly going green :smile:

**Comment by pcapriotti** [comment]
The golden tests haven't changed, because there are no changes in the JSON format. All I've done is wrap these maps in newtypes, since they have (and had already) special instances, but the corresponding JSON ignores the newtype.

The key decoding should be exactly the same as before (and hopefully the golden tests are helping to catch any possible mistakes there). Now the logic for converting keys to strings is in the `FromJSONKey` instance of `UserId`.

**Comment by jschaul** [comment]
Hm, at the moment we have a golden test for `UserClientMap Int`. In this PR, a change is made so the json returned on an endpoint is no longer `UserClientMap (Maybe Prekey)` but instead `UserClientPrekeyMap`. There is no golden test for that, and while I understand that logically this is a newtype wrapper change only, there currently is no backwards-compatibility-test here. Which is what we kind of want to do with the golden tests, no? Ensure that old JSON values sent by clients are still parsed as expected by the server, and new refactoring doesn't change the existing response.

I can go and change the definition of `UserClientPrekeyMap` to be `data UserClientPrekeyMap = UserClientPrekeyMap { foo = Text }` and no test would catch this backwards-compatibility problem. That's my concern. Please add a test :)

**Review by akshaymankar — APPROVED** [review]
Suggested minor changes. 

I am still not sure about having a special newtype for `Maybe Prekey`. I think it is fine to write a `ToSchema` instance for `Maybe a` as it has instances for `ToJSON`, `FromJSON` and `Swagger.ToSchema`. But I also see how this can be slightly dangerous, maybe we can transform most of our code to the `ToSchema` library and then write this instance and when someone needs `Maybe a` to not become a `Nothing` we can try to establish a newtype based pattern for that.

So, overall this looks good to me. Maybe let's wait for @jschaul's review too.

File: libs/wire-api/test/unit/Test/Wire/API/Golden/Manual/UserClientPrekeyMap.hs [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+  (mkUserClientPrekeyMap . fromList)
+    [ ((Id (fromJust (UUID.fromString "00000000-0000-0005-0000-000800000002"))), fromList []),
+      ((Id (fromJust (UUID.fromString "00000002-0000-0007-0000-000100000002"))), fromList []),
+      ( (Id (fromJust (UUID.fromString "00000002-0000-0008-0000-000300000006"))),
+        fromList [(ClientId {client = "12"}, Just testObject_Prekey_user_12)]
+      ),
+      ((Id (fromJust (UUID.fromString "00000002-0000-0008-0000-000300000008"))), fromList [])
+    ]
+
+testObject_UserClientPrekeyMap_20 :: UserClientPrekeyMap
```
[the remark this answers is not in the document]
Comment (pcapriotti) [reply libs/wire-api/test/unit/Test/Wire/API/Golden/Manual/UserClientPrekeyMap.hs:313]: Why? What's the harm? I've produced them by doing some replacements on the ones for `UserClientMap Int`.

File: libs/wire-api/src/Wire/API/User/Client.hs [hunk 2]
```diff
@@ -203,70 +206,95 @@ modelOtrClientMap = Doc.defineModel "OtrClientMap" $ do
...
+              (schemaDoc innerSchema ^. Swagger.schema . Swagger.example)
+          )
+
+newtype QualifiedUserClientPrekeyMap = QualifiedUserClientPrekeyMap
+  {getQualifiedUserClientPrekeyMap :: QualifiedUserClientMap (Maybe Prekey)}
+  deriving stock (Eq, Show)
+  deriving newtype (Arbitrary, Semigroup, Monoid)
+  deriving (FromJSON, ToJSON, Swagger.ToSchema) via Schema QualifiedUserClientPrekeyMap
+
+-- FUTUREWORK: wrap around another JSON object layer
```
[the remark this answers is not in the document]
Comment (pcapriotti) [reply libs/wire-api/src/Wire/API/User/Client.hs]: You're right, I just removed them.

File: libs/wire-api/test/unit/Test/Wire/API/Golden/Manual/UserClientPrekeyMap.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (akshaymankar) [reply libs/wire-api/test/unit/Test/Wire/API/Golden/Manual/UserClientPrekeyMap.hs:313]: When we generated them, there was no option but to generate a lot of them. When we're replacing them with Manual tests, I think we should try to limit them for saving on compile time and test run time.

**Review by jschaul — APPROVED** [review]

File: libs/wire-api/test/unit/Test/Wire/API/Golden/Manual/UserClientPrekeyMap.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pcapriotti) [reply libs/wire-api/test/unit/Test/Wire/API/Golden/Manual/UserClientPrekeyMap.hs:313]: Ok, that makes sense. Are you ok with leaving more than 2, though? They are used for the qualified part.

[the remark this answers is not in the document]
Comment (pcapriotti) [reply libs/wire-api/test/unit/Test/Wire/API/Golden/Manual/UserClientPrekeyMap.hs]: I left 8.

**Pull request merged**